pkg/reconciler/apis/permissionclaimlabel/permissionclaimlabel_reconcile.go

```diff
@@ -0,0 +1,248 @@
+
+	logger.V(6).Info("claim set details",
+		"expected", expectedClaims,
+		"unexpected", unexpectedClaims,
+		"toApply", needToApply,
+		"toRemove", needToRemove,
+		"all", allChanges,
+	)
+
+	var allErrs []error
+	applyErrors := sets.NewString()
+
+	for _, s := range allChanges.List() {
+		claim := claimFromSetKey(s)
+		claimLogger := logger.WithValues("claim", s)
+
+		informer, gvr, err := c.getInformerForGroupResource(claim.Group, claim.Resource)
+		if err != nil {
+			allErrs = append(allErrs, fmt.Errorf("error getting informer for group=%q, resource=%q: %w", claim.Group, claim.Resource, err))
+			if acceptedClaims.Has(s) {
+				applyErrors.Insert(s)
+			}
+			continue
+		}
+
+		claimLogger.V(4).Info("listing resources")
+		objs, err := informer.Informer().GetIndexer().ByIndex(indexers.ByLogicalCluster, clusterName.String())
+		if err != nil {
+			allErrs = append(allErrs, fmt.Errorf("error listing group=%q, resource=%q: %w", claim.Group, claim.Resource, err))
+			if acceptedClaims.Has(s) {
+				applyErrors.Insert(s)
+			}
+			continue
+		}
+
+		claimLogger.V(4).Info("got resources", "count", len(objs))
+
+		var claimErrs []error
+		for _, obj := range objs {
+			objectMeta, err := meta.Accessor(obj)
+			if err != nil {
+				claimErrs = append(claimErrs, fmt.Errorf("error getting meta accessor for %T: %w", obj, err))
+				continue
+			}
+
+			patchLogger := claimLogger.WithValues("objectNamespace", objectMeta.GetNamespace(), "objectName", objectMeta.GetName())
+			patchLogger.V(4).Info("patching to get claim labels updated")
+
+			// Empty patch, allowing the admission plugin to update the resource to the correct labels
+			err = c.patchGenericObject(ctx, objectMeta, gvr, clusterName)
+			if err != nil {
+				patchErr := fmt.Errorf("error patching %q %s|%s/%s: %w", gvr, clusterName, objectMeta.GetNamespace(), objectMeta.GetName(), err)
+				claimErrs = append(claimErrs, patchErr)
+				continue
+			}
+		}
+
+		if len(claimErrs) > 0 {
+			allErrs = append(allErrs, claimErrs...)
+
+			if acceptedClaims.Has(s) {
+				applyErrors.Insert(s)
+			}
+		}
+	}
+
+	var unexpectedOrInvalidErrors []error
+	for _, s := range unexpectedClaims.List() {
+		claim := claimFromSetKey(s)
+		unexpectedOrInvalidErrors = append(unexpectedOrInvalidErrors, fmt.Errorf("unexpected/invalid claim for %s.%s (identity %q)", claim.Resource, claim.Group, claim.IdentityHash))
+	}
+	if len(unexpectedOrInvalidErrors) > 0 {
+		conditions.MarkFalse(
+			apiBinding,
+			apisv1alpha1.PermissionClaimsValid,
+			apisv1alpha1.InvalidPermissionClaimsReason,
+			conditionsv1alpha1.ConditionSeverityError,
+			"Unexpected and/or invalid permission claims: %s",
+			aggregateerrors.NewAggregate(unexpectedOrInvalidErrors),
+		)
+	} else {
+		conditions.MarkTrue(apiBinding, apisv1alpha1.PermissionClaimsValid)
+	}
+
+	fullyApplied := expectedClaims.Difference(applyErrors)
+	apiBinding.Status.AppliedPermissionClaims = []apisv1alpha1.PermissionClaim{}
+	for _, s := range fullyApplied.List() {
```

Comment:
If this is large we can use `UnsortedList()`

Reply:
Fix in next commit